File: benchmarks/matmul_blocked/python/matmul_blocked.py
import sys

BLOCK = 32


def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 600

    a = [0.0] * (n * n)
    b = [0.0] * (n * n)
    for i in range(n):
        for j in range(n):
            a[i * n + j] = float((i * 3 + j) % 13)
            b[i * n + j] = float((i + j * 2) % 17)

    c = [0.0] * (n * n)
    for ii in range(0, n, BLOCK):
        i_max = min(ii + BLOCK, n)
        for kk in range(0, n, BLOCK):
            k_max = min(kk + BLOCK, n)
            for jj in range(0, n, BLOCK):
                j_max = min(jj + BLOCK, n)
                for i in range(ii, i_max):
                    for k in range(kk, k_max):
                        aik = a[i * n + k]
                        for j in range(jj, j_max):
                            c[i * n + j] += aik * b[k * n + j]

    expected = sum(a[k] * b[k * n] for k in range(n))
    assert c[0] == expected, "self-check failed: c[0][0] mismatch"

    print(f"{c[(n - 1) * n + (n - 1)]:.2f}")


if __name__ == "__main__":
    main()

File: benchmarks/matmul_blocked/rust/matmul_blocked.rs
use std::env;

const BLOCK: usize = 32;

fn main() {
    let n: usize = env::args().nth(1).map(|s| s.parse().unwrap()).unwrap_or(600);

    let mut a = vec![0.0f64; n * n];
    let mut b = vec![0.0f64; n * n];
    for i in 0..n {
        for j in 0..n {
            a[i * n + j] = ((i * 3 + j) % 13) as f64;
            b[i * n + j] = ((i + j * 2) % 17) as f64;
        }
    }

    let mut c = vec![0.0f64; n * n];
    let mut ii = 0;
    while ii < n {
        let i_max = (ii + BLOCK).min(n);
        let mut kk = 0;
        while kk < n {
            let k_max = (kk + BLOCK).min(n);
            let mut jj = 0;
            while jj < n {
                let j_max = (jj + BLOCK).min(n);
                for i in ii..i_max {
                    for k in kk..k_max {
                        let aik = a[i * n + k];
                        for j in jj..j_max {
                            c[i * n + j] += aik * b[k * n + j];
                        }
                    }
                }
                jj += BLOCK;
            }
            kk += BLOCK;
        }
        ii += BLOCK;
    }

    let expected: f64 = (0..n).map(|k| a[k] * b[k * n]).sum();
    assert_eq!(c[0], expected, "self-check failed: c[0][0] mismatch");

    println!("{:.2}", c[(n - 1) * n + (n - 1)]);
}
